gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java

```diff
@@ -191,59 +197,168 @@ private String dependencyPathsOfProblematicJars(
+  private static class ResolvedComponentResultNode {
+    ResolvedComponentResult componentResult;
+
+    ResolvedComponentResultNode parent;
+
+    ResolvedComponentResultNode(
+        ResolvedComponentResult componentResult, ResolvedComponentResultNode parent) {
+      this.componentResult = componentResult;
+      this.parent = parent;
+    }
+
+    boolean hasParent(ResolvedComponentResult other) {
+      if (componentResult.equals(other)) {
+        return true;
+      }
+      if (parent == null) {
+        return false;
+      }
+      return parent.hasParent(other);
+    }
+
+    public String pathFromRoot() {
+      ResolvedComponentResultNode iter = this;
+      List<String> dependencyPathElementsReversed = new ArrayList<>();
+      while (iter != null) {
+        dependencyPathElementsReversed.add(formatComponentResult(iter.componentResult));
+        iter = iter.parent;
+      }
+      Collections.reverse(dependencyPathElementsReversed);
+      String dependencyPath = Joiner.on(" / ").join(dependencyPathElementsReversed);
+      return dependencyPath;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("componentResult", componentResult)
+          .add("parent", parent)
+          .toString();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (this == other) {
+        return true;
+      }
+      if (other == null || getClass() != other.getClass()) {
+        return false;
+      }
+      ResolvedComponentResultNode that = (ResolvedComponentResultNode) other;
+      return Objects.equals(componentResult, that.componentResult)
+          && Objects.equals(parent, that.parent);
     }
 
-    for (DependencyResult dependencyResult : item.getDependencies()) {
-      if (dependencyResult instanceof ResolvedDependencyResult) {
-        ResolvedDependencyResult resolvedDependencyResult =
-            (ResolvedDependencyResult) dependencyResult;
-        ResolvedComponentResult child = resolvedDependencyResult.getSelected();
+    @Override
+    public int hashCode() {
+      return Objects.hash(componentResult, parent);
+    }
+  }
+
+  /**
+   * Returns mapping from Maven coordinates to their dependency paths appearing in the dependency
+   * graph
+   *
+   * @param componentResult The root project in the dependency graph
+   * @param targetCoordinatesSet The Maven coordinates to check their dependency paths
+   */
+  private ListMultimap<String, String> groupCoordinatesToDependencyPaths(
+      ResolvedComponentResult componentResult, Set<String> targetCoordinatesSet) {
+
+    ListMultimap<String, String> coordinatesToDependencyPaths =
+        MultimapBuilder.hashKeys().arrayListValues().build();
+    // No need to print the same circular dependency multiple times
+    Set<ResolvedComponentResult> checkedCircularDependency = new HashSet<>();
+
+    for (String targetCoordinates : targetCoordinatesSet) {
```

Comment:
This loop is new and may increase the runtime by the size of targetCoordinatesSet. This should not be a problem as this PR is eliminating traversing duplicate paths which would increase the operation in an exponential order.

Reply:
The runtime speed was not the problem. The problem of OutOfMemoryError (https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/2187) seems to be caused by a number of paths stored in the multimap. Omitting the duplicate paths will solve the problem.